plugins/auth/irods_auth_plugin-native/libirods_auth_plugin-native.cpp

```diff
@@ -0,0 +1,467 @@
+#include "authentication_plugin_framework.hpp"
+
+#include "base64.h"
+
+int get64RandomBytes( char *buf );
+void setSessionSignatureClientside( char* _sig );
+void _rsSetAuthRequestGetChallenge( const char* _c );
+
+namespace
+{
+    using json = nlohmann::json;
+
+    auto request(rcComm_t* comm, const json& msg)
+    {
+        auto str = msg.dump();
+
+        bytesBuf_t inp;
+        inp.buf = (void*)str.c_str();
+        inp.len = str.size();
+
+        bytesBuf_t* resp{};
+        auto err = procApiRequest(comm, 110000, (void*)&inp, NULL, (void**)&resp, NULL );
+
+        if (err < 0) {
+            THROW(err, "failed to perform request");
+        }
+
+        return json::parse((char*)resp->buf, (char*)resp->buf + resp->len);
+    } // request
+
+    // TODO: std::vector could be a const ref
+    auto verify(const json& p, std::vector<std::string> n)
+    {
+        for(auto&& x : n) {
+            if(!p.contains(x)) {
+                return std::make_tuple(false, x);
+            }
+        }
+
+        return std::make_tuple(true, std::string{});
+    } // verify
+
+    template<typename T>
+    auto get(const std::string& n, const json& p)
+    {
+        if(!p.contains(n)) {
+            THROW(SYS_INVALID_INPUT_PARAM, fmt::format(
+                  "authentication request missing [{}] parameter", n));
+        }
+
+        return p.at(n).get<T>();
+    } // get
+
+#ifdef RODS_SERVER
+    void throw_on_insufficient_privilege_for_proxy_user(const rsComm_t* const _comm,
+                                                        int _proxy_priv)
+    {
+        // If the proxy user is a rodsadmin in the local zone, it can do whatever it wants.
+        if (_proxy_priv >= LOCAL_PRIV_USER_AUTH) {
+            return;
+        }
+
+        // A user is allowed to authenticate itself.
+        if (std::string_view{_comm->proxyUser.userName} == _comm->clientUser.userName) {
+            return;
+        }
+
+        // A proxy user with remote priviliged user authorization can only authenticate client
+        // users from the same zone.
+        if (_proxy_priv >= REMOTE_PRIV_USER_AUTH &&
+            std::string_view{_comm->proxyUser.rodsZone} == _comm->clientUser.rodsZone) {
+            return;
+        }
+
+        // This is a proxy user attempting to authenticate a client user with insufficient
+        // permissions to do so.
+        THROW(SYS_PROXYUSER_NO_PRIV, fmt::format(
+              "proxy user: [{}] with {} no priv to auth client user: [{}].",
```

Comment:
this statement is a little confusing. how about:

"Proxy user [{}] is not allowed to authenticate client user [{}]."

i'm not sure how to include the privilege level. is it needed?

Reply:
Perhaps something like "Requires auth level [{}] but proxy user has [{}]." ?

Reply:
I'll just exclude the privilege lever. I don't think we need it in this message.

Reply:
having a log message is good. i think we should leave the privilege level out for now. we can always revisit it if it becomes a problem.

Reply:
Agreed